optimize-onnx/src/optimize.cpp

```diff
@@ -0,0 +1,176 @@
+#include <queue>
+
+#include "optimize.h"
+
+namespace onnx { namespace optimization {
+
+bool isPureOperator(Node * n) {
+  // TODO This is a lie.
+  return true;
+}
+
+// Split the graph into 'init' and 'predict' nets. This is kind of
+// like constant folding, except that rather than actually execute the
+// constant computations, we simply split them out into a separate
+// graph. Nodes that have any transitive dependency on the
+// initializers, or on impure operators, must remain in the predict
+// net. All others may be moved to the init net.
+//
+// This function destructively mutates the graph into either the init
+// or the predict net. If you want both, which you probably do,
+// arrange to call it twice.
+//
+// NOTE POTENTIAL BREAKAGE:
+//
+// The ONNX spec provides no guarantees about "staging", i.e. which
+// inputs change on every invocation vs which generally stay the same.
+// Here we make the assumption that inputs which have an initializer
+// value provided for them vary only between invocations of the init
+// net, and are constant across runs of the predict net.
+//
+std::shared_ptr<Graph> split_init_and_predict(std::shared_ptr<Graph> g, bool init, bool predict) {
+  // The first step is a graph traversal to identify what is reachable
+  // from the mutable intializers. Since we have a bipartite graph, we
+  // must track both Nodes and Values that we have reached (as either
+  // could be terminal objects in the graph), but it's sufficient to
+  // maintain a frontier consisting only of Values.
+  std::unordered_set<Value *> live_values;
+  std::unordered_set<Node *> live_nodes;
+
+  {
+    std::queue<Value *> frontier;
+
+    {
+      std::unordered_set<std::string> initializer_names(
+        g->initializer_names().begin(),
+        g->initializer_names().end());
+
+      // Seed frontier with graph input Values that do not have a
+      // corresponding Initializer.
+      for (Value * v : g->inputs()) {
+        if (initializer_names.find(v->uniqueName()) == initializer_names.end()) {
+          live_values.insert(v);
+          frontier.push(v);
+        }
+      }
+    }
+
+    // Additionally seed frontier with the output Values of impure
+    // Nodes.
+    for (Node * n : g->nodes()) {
+      if (isPureOperator(n)) {
+        continue;
+      }
+      live_nodes.insert(n);
+      for (Value * v : n->outputs()) {
+        live_values.insert(v);
+        frontier.push(v);
+      }
+    }
+
+    // Mark as live all Values and Nodes that are reachable from the
+    // seed Values and Nodes.
+    while (!frontier.empty()) {
+      for (Use u : frontier.front()->uses()) {
+        if (live_nodes.find(u.user) != live_nodes.end()) {
```

Comment:
`live_nodes.count(u.user) > 0` is slightly less verbose but does the same thing.